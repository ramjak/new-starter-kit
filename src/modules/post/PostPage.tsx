import React, { useCallback } from "react";
import { Button } from "@material-ui/core";
import { Add } from "@material-ui/icons";
import usePost from "../../domainHooks/usePost";
import styles from "./PostPage.module.scss";
import SinglePost from "./SinglePost";
import ROUTES, { useNavigateTo } from "../../routes";

interface IPostPage {}

function PostPage(props: IPostPage) {
  const { data, destroy } = usePost();
  const navigateTo = useNavigateTo();

  // useEffect(() => {
  //   const currentTitle = document.title;
  //   document.title = 'PostPage';
  //   return () => { document.title = currentTitle };
  // }, []);

  const goToCreatePostPage = useCallback(() => {
    navigateTo(ROUTES.createPost);
  }, [navigateTo]);

  return (
    <main className={styles.postPage}>
      <h1>Post List</h1>
      <Button
        variant="contained"
        color="default"
        className={styles.addPostBtn}
        onClick={goToCreatePostPage}
      >
        <Add />
      </Button>
      <hr />
      {data.slice(0, 5).map((item) => (
        <SinglePost key={item.id} item={item} destroy={destroy} />
      ))}
    </main>
  );
}

export default PostPage;
