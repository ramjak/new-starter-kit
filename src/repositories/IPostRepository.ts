import IPost, { IRawPost } from "../domains/post";

export default interface IPostRepository {
  store(payload: IRawPost): Promise<unknown>;
  read(id: string): Promise<IPost>;
  update(payload: IRawPost, id: string): Promise<unknown>;
  delete(id: string): Promise<unknown>;
  getAll(): Promise<IPost[]>;
}
