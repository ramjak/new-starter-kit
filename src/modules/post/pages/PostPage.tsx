import { useCallback, useEffect, useState } from "react";
import { Button } from "@material-ui/core";
import { Add } from "@material-ui/icons";
import { useInjection } from "inversify-react";
import styles from "./PostPage.module.scss";
import SinglePost from "../components/SinglePost";
import ROUTES, { IRoute, useNavigateTo as useNT } from "../../../routes";
import TYPES from "../../../services/types";
import IPostRepository from "../../../repositories/IPostRepository";
import IPost from "../../../domains/post";

interface IPostPage {
  // eslint-disable-next-line react/no-unused-prop-types
  useNavigateTo?: () => (route: IRoute, ...params: string[] | number[]) => void;
}

const PostPage = (props: IPostPage) => {
  const { useNavigateTo } = { useNavigateTo: useNT, ...props };
  const postRepository = useInjection<IPostRepository>(TYPES.PostRepository);
  const [posts, setPosts] = useState<IPost[]>([]);
  const navigateTo = useNavigateTo();

  const loadPosts = useCallback(async () => {
    const loadedPosts = await postRepository.getAll();
    setPosts(loadedPosts);
  }, [postRepository]);

  const deletePost = useCallback((id) => postRepository.delete(id), [
    postRepository,
  ]);

  useEffect(() => {
    loadPosts().catch((e) => console.log(e));
  }, [loadPosts]);

  const navigateToSinglePost = useCallback(
    (id) => navigateTo(ROUTES.viewSinglePost, id),
    [navigateTo]
  );

  const navigateToEditForm = useCallback(
    (id) => navigateTo(ROUTES.editSinglePost, id),
    [navigateTo]
  );

  const goToCreatePostPage = useCallback(() => {
    navigateTo(ROUTES.createPost);
  }, [navigateTo]);

  return (
    <main className={styles.postPage}>
      <h1>Post List</h1>
      <Button
        variant="contained"
        color="default"
        className={styles.addPostBtn}
        onClick={goToCreatePostPage}
      >
        <Add />
      </Button>
      <hr />
      {posts.slice(0, 5).map((item) => (
        <SinglePost
          key={item.id}
          item={item}
          view={navigateToSinglePost}
          edit={navigateToEditForm}
          destroy={deletePost}
        />
      ))}
    </main>
  );
};

export default PostPage;
