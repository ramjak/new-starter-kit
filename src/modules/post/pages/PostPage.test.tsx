import "reflect-metadata";
import React from "react";
import { Provider } from "inversify-react";
import { Container } from "inversify";
import renderer from "react-test-renderer";
import { render, waitFor, screen, getByTestId } from "@testing-library/react";
import TYPES from "../../../services/types";
import IPostRepository from "../../../repositories/IPostRepository";
import PostPage from "./PostPage";
import IPost from "../../../domains/post";

describe("PostPage", () => {
  const samplePosts: IPost[] = [
    { title: "lol", id: 1, body: "test1", userId: 1 },
    { title: "title", id: 2, body: "test2", userId: 2 },
  ];

  const container = new Container();
  container.bind<IPostRepository>(TYPES.PostRepository).toConstantValue({
    getAll: () => {
      return Promise.resolve(samplePosts);
    },
    read: jest.fn(),
    store: jest.fn(),
    update: jest.fn(),
    delete: jest.fn(),
  });

  it("renders base structure correctly at first load", () => {
    const result = renderer
      .create(
        <Provider container={container}>
          <PostPage />
        </Provider>
      )
      .toJSON();

    expect(result).toMatchSnapshot();
  });

  // todo: isolate the page from the the children components
  it("render fetched posts", async () => {
    render(
      <Provider container={container}>
        <PostPage />
      </Provider>
    );

    await waitFor(() => screen.getByRole("heading"));
    const postArticles = screen.getAllByRole("article");

    expect(getByTestId(postArticles[0], "view-btn")).toHaveTextContent("lol");
    expect(getByTestId(postArticles[0], "article-body")).toHaveTextContent(
      "test1"
    );

    expect(getByTestId(postArticles[1], "view-btn")).toHaveTextContent("title");
    expect(getByTestId(postArticles[1], "article-body")).toHaveTextContent(
      "test2"
    );
  });
});
