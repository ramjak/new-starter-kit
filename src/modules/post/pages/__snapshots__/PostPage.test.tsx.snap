// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostPage renders base structure correctly at first load 1`] = `
<main
  className="postPage"
>
  <h1>
    Post List
  </h1>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained addPostBtn"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </span>
  </button>
  <hr />
</main>
`;
